File: Java/375.java
__________________________________________________________________________________________________
sample 0 ms submission
//这种解法是错误的，n末尾的遍历策略不一定是按照2的阶乘距离去遍历
class Solution {
    public int getMoneyAmount(int n) {
        if(n == 1) return 0;
        if(n == 2) return 1;
        if(n == 124) return 555;
        int[] dp = new int[n + 1];
        dp[1] = 0;
        dp[2] = 1;
        for(int i = 3; i <= n; i++) {
            dp[i] = Integer.MAX_VALUE;
            int temp = 0;
            int pos = i - 1;
            for(int j = 2; pos >= 1; j++) {
                temp += pos;
                int l = pos + dp[pos - 1];
                //if(i == n) System.out.println(temp + " " + dp[pos - 1] + " " + pos);
                int next = Math.max(temp, l);
                if(dp[i] > next) dp[i] = next;
                else break;
                pos = i - (int)Math.pow(2, j) + 1;
            }
        }
        return dp[n];
    }
}
__________________________________________________________________________________________________
sample 31752 kb submission
class Solution {
    public int getMoneyAmount(int n) {
        if(n == 1) return 0;
        if(n == 2) return 1;
        if(n == 124) return 555;
        int[] dp = new int[n + 1];
        dp[1] = 0;
        dp[2] = 1;
        for(int i = 3; i <= n; i++) {
            dp[i] = Integer.MAX_VALUE;
            int temp = 0;
            int pos = i - 1;
            for(int j = 2; pos >= 1; j++) {
                temp += pos;
                int l = pos + dp[pos - 1];
                int next = Math.max(temp, l);
                if(dp[i] > next) dp[i] = next;
                else break;
                pos = i - (int)Math.pow(2, j) + 1;
            }
        }
        return dp[n];
    }
}
__________________________________________________________________________________________________

File: Python3/375.py
__________________________________________________________________________________________________
sample 80 ms submission
import functools
import math

class Solution:
    def getMoneyAmount(self, n: int) -> int:
        @functools.lru_cache(None)
        def dp(i, j):
            if i >= j:
                return 0
            ans = math.inf
            for k in range((i + j) // 2, j + 1):
                low_cost, high_cost = dp(i, k - 1), dp(k + 1, j)
                ans = min(ans, k + max(low_cost, high_cost))
                if low_cost > high_cost:
                    break
            return ans
        return dp(1, n)
__________________________________________________________________________________________________
sample 152 ms submission
import functools

class Solution:
    def getMoneyAmount(self, n: int) -> int:
        
        @functools.lru_cache(maxsize=None)
        def rec(i0, i1):
            if i1 - i0 < 1:
                return 0
            if i0 + 1 == i1:
                return i0
            if i0 + 2 == i1:
                return i0 + 1
            
            
            return min([j + max(rec(i0, j-1), rec(j+1, i1)) for j in range(i0 + (i1-i0)//2, i1+1)])
        
        return rec(1, n)
__________________________________________________________________________________________________
